File: athena_ts/packages/athena-research/src/supervisor/ranker.ts
/** Hypothesis scoring policy, ranking and selection-local deduplication. */

import type { Hypothesis, ResearchTree } from "@athena/core"

/** 候选相对其冻结参照的可信结果。 */
export const Outcome = {
  WIN: "WIN",
  DRAW: "DRAW",
  LOSS: "LOSS",
} as const
export type Outcome = (typeof Outcome)[keyof typeof Outcome]

/** Supervisor 调度器使用的窄策略。 */
export interface HypothesisPolicy {
  seed(parent: Hypothesis | null): number
  priority(hypothesis: Hypothesis): number
  settle(referencePriority: number, outcome: Outcome): number
}

/**
 * 单侧 Elo 更新，固定 0.5 期望得分。
 *
 * TODO(search-policy): 有足够真实搜索轨迹后，用 evidence-aware 调度策略替换
 * EloPolicy；Supervisor 只依赖 HypothesisPolicy。
 */
export class EloPolicy implements HypothesisPolicy {
  static readonly ROOT_PRIORITY = 1000.0

  constructor(private readonly k: number = 32.0) {
    if (!Number.isFinite(k) || k <= 0) {
      throw new Error("Elo k must be a positive finite number")
    }
  }

  seed(parent: Hypothesis | null): number {
    return parent === null ? EloPolicy.ROOT_PRIORITY : parent.priority
  }

  priority(hypothesis: Hypothesis): number {
    return hypothesis.priority
  }

  settle(referencePriority: number, outcome: Outcome): number {
    if (!Number.isFinite(referencePriority)) {
      throw new Error("reference priority must be finite")
    }
    const score =
      outcome === Outcome.WIN ? 1.0 : outcome === Outcome.DRAW ? 0.5 : 0.0
    return referencePriority + this.k * (score - 0.5)
  }
}

export function tokenize(text: string): Set<string> {
  return new Set(text.toLowerCase().match(/[a-z0-9]+/g) ?? [])
}

export function jaccard(left: Set<string>, right: Set<string>): number {
  if (left.size === 0 && right.size === 0) return 0.0
  let intersection = 0
  for (const token of left) {
    if (right.has(token)) intersection += 1
  }
  return intersection / (left.size + right.size - intersection)
}

function tokens(hypothesis: Hypothesis): Set<string> {
  return tokenize(hypothesis.statement + " " + hypothesis.intervention)
}

/** Evidence and specificity provide the unchanged [0.4, 1.0] cold-start prior. */
function rubricPrior(hypothesis: Hypothesis): number {
  const evidence = hypothesis.sources.length > 0 ? 1.0 : 0.0
  const specific = tokenize(hypothesis.intervention).size >= 3 ? 1.0 : 0.0
  return 0.4 + 0.3 * evidence + 0.3 * specific
}

/** Greedy selection: input order wins ties; neither input is mutated. */
export function deduplicate(
  candidates: Hypothesis[],
  existing: Hypothesis[],
  threshold: number
): Hypothesis[] {
  const kept: Hypothesis[] = []
  const seen = existing.map(tokens)
  for (const candidate of candidates) {
    const words = tokens(candidate)
    if (seen.some((other) => jaccard(words, other) >= threshold)) continue
    kept.push(candidate)
    seen.push(words)
  }
  return kept
}

export interface RankConfig {
  priorWeight: number
  strengthWeight: number
  noveltyWeight: number
  costWeight: number
  dedupThreshold: number
}

export const DEFAULT_RANK_CONFIG: RankConfig = {
  priorWeight: 0.4,
  strengthWeight: 0.3,
  noveltyWeight: 0.2,
  costWeight: 0.1,
  dedupThreshold: 0.8,
}

/** Rank by prior + normalized policy strength + novelty - clipped cost. */
export class Selector {
  constructor(
    private readonly policy: Pick<HypothesisPolicy, "priority">,
    private readonly config: RankConfig = DEFAULT_RANK_CONFIG
  ) {
    const weights = [config.priorWeight, config.strengthWeight, config.noveltyWeight]
    if (weights.some((weight) => !Number.isFinite(weight) || weight < 0 || weight > 1)) {
      throw new Error("selector weights must be between 0 and 1")
    }
    if (!Number.isFinite(config.costWeight) || config.costWeight < 0) {
      throw new Error("cost_weight must be finite and nonnegative")
    }
    if (!Number.isFinite(config.dedupThreshold) || config.dedupThreshold <= 0 || config.dedupThreshold > 1) {
      throw new Error("dedup_threshold must be in (0, 1]")
    }
  }

  deduplicate(candidates: Hypothesis[], existing: Hypothesis[]): Hypothesis[] {
    return deduplicate(candidates, existing, this.config.dedupThreshold)
  }

  rank(tree: ResearchTree, candidates: Hypothesis[]): Hypothesis[] {
    const priorities = new Map<string, number>()
    for (const hypothesis of candidates) {
      if (hypothesis.id !== null) {
        priorities.set(hypothesis.id, this.policy.priority(hypothesis))
      }
    }
    const values = [...priorities.values()].filter(Number.isFinite)
    const low = values.length > 0 ? Math.min(...values) : 0.0
    const high = values.length > 0 ? Math.max(...values) : 0.0
    const span = high - low
    const settled = Object.values(tree.toDict().hypotheses)
      .filter((hypothesis) => hypothesis.status !== "PROPOSED")
      .map(tokens)
    const config = this.config
    const scored = candidates.map((hypothesis) => {
      const strength = hypothesis.id === null || span <= 0
        ? 0.5
        : (priorities.get(hypothesis.id)! - low) / span
      const words = tokens(hypothesis)
      let similarity = 0
      for (const other of settled) similarity = Math.max(similarity, jaccard(words, other))
      const cost = Math.min(Math.max(hypothesis.cost, 0.0), 1.0)
      const score =
        config.priorWeight * rubricPrior(hypothesis) +
        config.strengthWeight * strength +
        config.noveltyWeight * (1 - similarity) -
        config.costWeight * cost
      return { hypothesis, score }
    })
    scored.sort((left, right) => {
      if (right.score !== left.score) return right.score - left.score
      return (left.hypothesis.order ?? 0) - (right.hypothesis.order ?? 0)
    })
    return scored.map(({ hypothesis }) => hypothesis)
  }
}

File: src/athena/research/supervisor/ranker.py
"""Hypothesis ranking and deduplication for rolling SEARCH.

Replaces the pure ``(-priority, order)`` FIFO sort with a small UCB-style
selector: a rubric cold-start prior + accumulated strength (a Bradley-Terry
proxy from the policy's real WIN/DRAW/LOSS rating) + a novelty exploration
bonus - a normalized cost penalty. All signals are deterministic and pure
functions of the tree, so they stay testable and reproducible.
"""

import math
import re

from pydantic import BaseModel, ConfigDict, Field

from athena.core.research_models import Hypothesis
from athena.core.research_tree import ResearchTree

_WORD = re.compile(r"[a-z0-9]+")


def tokenize(text: str) -> set[str]:
    """Lowercase word tokens for a cheap, deterministic similarity."""
    return set(_WORD.findall(text.lower()))


def jaccard(left: set[str], right: set[str]) -> float:
    """Jaccard similarity; an empty union yields 0.0."""
    if not left and not right:
        return 0.0
    return len(left & right) / len(left | right)


def _text_similarity(left: Hypothesis, right: Hypothesis) -> float:
    """Similarity over statement + intervention tokens (the dedup key)."""
    a = tokenize(left.statement) | tokenize(left.intervention)
    b = tokenize(right.statement) | tokenize(right.intervention)
    return jaccard(a, b)


def rubric_prior(hypothesis: Hypothesis, tree: ResearchTree) -> float:
    """Cold-start prior in [0.4, 1.0] from specificity alone.

    A pending hypothesis has no experiment data yet, so this rubric is the only
    reliable signal. It is deliberately small and transparent; a future
    ReflectionAgent rubric can replace it with richer scores.

    **有引用不再加分。** 早先这里给非空 ``sources`` 加 0.3（占总分 0.12），意图是让有据
    可依的假设先跑。真机（2026-08-16 第 12 次）表明它买到的是装饰而不是依据：带引用与
    不带引用的假设提的是同一批干预，引用是事后贴上去的，且经常张冠李戴——一篇《数据增强
    综述》被用来支持"两两交互特征"。奖励"有没有引用"就是在为贴标签付钱。

    引用本身仍然有价值（可追溯、可复核），只是不该换算成优先级。真要让文献影响排序，得
    先有"这条引用确实支持这个主张"的判据，而那还不存在。
    """
    del tree
    specific = 1.0 if len(tokenize(hypothesis.intervention)) >= 3 else 0.0
    return 0.4 + 0.6 * specific


def novelty(hypothesis: Hypothesis, tree: ResearchTree) -> float:
    """1 - max similarity to settled hypotheses (1.0 when nothing is settled).

    This is the UCB exploration term: it rewards directions unlike anything
    already tested, instead of repeating near-duplicate interventions.
    """
    settled = [item for item in tree.hypotheses() if item.status != "PROPOSED"]
    if not settled:
        return 1.0
    return 1.0 - max(_text_similarity(hypothesis, other) for other in settled)


def deduplicate(
    candidates: list[Hypothesis],
    existing: list[Hypothesis],
    threshold: float,
) -> list[Hypothesis]:
    """Keep candidates distinct from ``existing`` and from each other.

    Greedy in input order (earlier candidates win ties). ``existing`` should be
    every hypothesis already in the graph, regardless of status.
    """
    kept: list[Hypothesis] = []
    seen: list[Hypothesis] = list(existing)
    for candidate in candidates:
        if any(_text_similarity(candidate, other) >= threshold for other in seen):
            continue
        kept.append(candidate)
        seen.append(candidate)
    return kept


class RankConfig(BaseModel):
    """Weights for the selector's deterministic score."""

    model_config = ConfigDict(frozen=True, extra="forbid")

    prior_weight: float = Field(default=0.4, ge=0.0, le=1.0)
    strength_weight: float = Field(default=0.3, ge=0.0, le=1.0)
    novelty_weight: float = Field(default=0.2, ge=0.0, le=1.0)
    cost_weight: float = Field(default=0.1, ge=0.0)
    dedup_threshold: float = Field(default=0.8, gt=0.0, le=1.0)


class Selector:
    """Rank pending hypotheses by ``prior + strength + novelty - cost``.

    ``strength`` is the policy's accumulated rating — the Bradley-Terry proxy
    fed by real experiment WIN/DRAW/LOSS outcomes; ``novelty`` is the UCB
    exploration bonus; ``cost`` is a normalized penalty for expensive work.
    """

    def __init__(self, policy, config: RankConfig | None = None) -> None:
        self._policy = policy
        self._config = config or RankConfig()

    @property
    def dedup_threshold(self) -> float:
        """Threshold shared by ranking-time dedup callers."""
        return self._config.dedup_threshold

    def deduplicate(
        self, candidates: list[Hypothesis], existing: list[Hypothesis]
    ) -> list[Hypothesis]:
        """Drop candidates too similar to the graph or to each other."""
        return deduplicate(candidates, existing, self._config.dedup_threshold)

    def rank(
        self, tree: ResearchTree, candidates: list[Hypothesis]
    ) -> list[Hypothesis]:
        """Return candidates sorted best-first, tie-broken by ``order``."""
        strengths = self._normalized_strengths(candidates, tree)
        return sorted(
            candidates,
            key=lambda h: (-self._score(h, tree, strengths), h.order or 0),
        )

    def _normalized_strengths(
        self, candidates: list[Hypothesis], tree: ResearchTree
    ) -> dict[str | None, float]:
        priorities = {
            hypothesis.id: self._policy.priority(hypothesis, tree)
            for hypothesis in candidates
            if hypothesis.id is not None
        }
        values = [value for value in priorities.values() if math.isfinite(value)]
        low = min(values) if values else 0.0
        high = max(values) if values else 0.0
        span = high - low
        if span <= 0:
            return {key: 0.5 for key in priorities}
        return {key: (value - low) / span for key, value in priorities.items()}

    def _score(
        self,
        hypothesis: Hypothesis,
        tree: ResearchTree,
        strengths: dict[str | None, float],
    ) -> float:
        config = self._config
        prior = rubric_prior(hypothesis, tree)
        strength = strengths.get(hypothesis.id, 0.5)
        explore = novelty(hypothesis, tree)
        cost = min(max(hypothesis.cost, 0.0), 1.0)
        return (
            config.prior_weight * prior
            + config.strength_weight * strength
            + config.novelty_weight * explore
            - config.cost_weight * cost
        )
